portal/src/ErrorBoundSuspense.tsx

```diff
@@ -0,0 +1,159 @@
+import React, {
+  Suspense,
+  SuspenseProps,
+  lazy,
+  useCallback,
+  useMemo,
+  useState,
+  useEffect,
+} from "react";
+import { ErrorBoundary, type ErrorBoundaryProps } from "react-error-boundary";
+
+export interface ErrorBoundSuspenseProps {
+  errorBoundaryFallbackComponent: NonNullable<
+    ErrorBoundaryProps["FallbackComponent"]
+  >;
+  suspenseFallback: SuspenseProps["fallback"];
+  factory: () => Promise<{ default: React.ComponentType<any> }>;
+  children: (Component: React.ComponentType<any>) => React.ReactNode;
+}
+
+const extractImportArgRegex = /import\(["']([-_./a-zA-Z0-9]+)["']\)/;
+
+function extractURLToModuleFromSourceCode(
+  factory: () => Promise<{ default: React.ComponentType<any> }>
+): URL | null {
+  // The sourceCode contains a relative URL.
+  // It is relative to the original module that loads it.
+  // For example, the original module that loads it may be
+  // https://portal/shared-assets/index-deadbeef.js
+  // The relative URL is "./SomeScreen-deadbeef.js"
+  //
+  // If we call import directly on the relative URL,
+  // it is relative to window.location, that is https://portal/project/PROJECT_ID/configuration/authentication
+  // Thus load https://portal/project/PROJECT_ID/configuration/authentcation/SomeScreen-deadbeef.js
+  // That will give a index.html because we are a SPA.
+  //
+  // To get the absolute URL to the module,
+  // we have to know the URL to the original module.
+  // This is impossible because we cannot know it from here.
+  //
+  // Fortunately, we can make an assumption here.
+  // Since all assets are put in the assets directory,
+  // if we know the URL to the current module (that is, this file),
+  // then we can assume that the original module is next to this file.
+
+  try {
+    // We are doing some arcane art here.
+    // To know the URL to load the module,
+    // we call toString on factory.
+    // That is, we peek the compiled source code of the function!
+    const sourceCode = factory.toString();
+    const match = extractImportArgRegex.exec(sourceCode);
+    if (match != null) {
+      const relativePath = match[1];
+      const urlToThisFile = import.meta.url;
+      // new URL("./SomeScreen.js", "https://localhost/shared-assets/index.js") =>
+      // https://localhost/shared-assets/SomeScreen.js
+      const url = new URL(relativePath, urlToThisFile);
+      return url;
+    }
+  } catch (_e: unknown) {
+    // ignore
+  }
+
+  return null;
+}
+
+function makeFactory(urlToModule: URL) {
+  // Browsers cache modules, including failed attempt.
```

Comment:
🤔 It seems we should add "no-cache" to 404 responses of our server?

Reply:
The stdlib does that already. See https://pkg.go.dev/net/http#ServeContent But browser seem ignore HTTP cache header in this particular case.

Reply:
https://github.com/user-attachments/assets/16a8d895-396e-4938-a17e-6976f43ac331

The source code of the server

```golang
package main

import (
	"fmt"
	"net/http"
	"net/http/httputil"
)

var html = `<!DOCTYPE html>
<html>
<head>
	<title>Hello</title>
</head>
<body>
</body>
</html>
`

func main() {
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte(html))
	})
	http.HandleFunc("/index.js", func(w http.ResponseWriter, r *http.Request) {
		b, _ := httputil.DumpRequest(r, true)
		fmt.Printf("%v\n", string(b))
		w.Header().Set("Cache-Control", "no-store")
		w.WriteHeader(500)
	})
	http.ListenAndServe(":3000", nil)
}

```